### PR TITLE
Intersection

### PR DESCRIPTION
**Pull request by brettz9**
Fixes part of #50.

feat: add intersection type (typescript and permissive only for now until jsdoc/closure may accept)

Note that I did not do an exhaustive analysis of where intersections may occur or what its operands may be, and merely copied the types for unions. (I would have thought that intersections might prevent string/number literals, but the TS playground does not seem to have an issue with it.)

### CODE REVIEW

**Comment by sandersn** [comment]
> I would have thought that intersections might prevent string/number literals

Do you mean `1 & 2` or `"hello" & "goodbye"`? Vacuous intersections collapse to `never`, the bottom type, which is only usable to indicate errors. But intersecting two types is never an error; it may just create a type that's not practically usable.

File: lib/publishing.js [hunk 1]
```diff
@@ -61,6 +61,10 @@ function createDefaultPublisher() {
       return format('%s|%s', concretePublish(unionNode.left),
                     concretePublish(unionNode.right));
     },
+    INTERSECTION (unionNode, concretePublish) {
```
Comment (sandersn) [inline lib/publishing.js:64]: I would prefer to format both intersections and unions with whitespace.

Comment (brettz9) [reply lib/publishing.js:64]: Agreed. Added!

**Comment by brettz9** [comment]
Btw, I added a link to the TS intersection links, but note that the TS docs on JSDoc only explicitly mention support for unions, not intersections (they mention supporting any TS type, but it doesn't say any TS operator). However, it seems from comments others have made (and the issue requesting adding support), that they are supported within jsdoc.

I also added a link to the jsdoc issue filed that is requesting Closure support but surprisingly could find no issue requesting support for intersections in Closure (and Closure docs do not mention it as a supported operator and searching source seems to only turn up ampersands for the JS bitwise ampersand), so it seems reasonable to keep excluding it for closure mode.

**Comment by sandersn** [comment]
> the TS docs on JSDoc only explicitly mention support for unions, not intersections (they mention supporting any TS type, but it doesn't say any TS operator)
I think that's a case of not explaining our terminology enough in the docs, then, because we consider type operators, like `&`, to create new type kinds, like intersections. The JSDoc page should probably link to the types/advanced types pages as well.

**Comment by brettz9** [comment]
Sure, yeah I figured it may be considering them as types, but a little thrown off with unions being specifically called out (and not seeming like just as an example).

**Pull request merged**